### DISCUSSION TITLE
qemu-arm-static missing

### DISCUSSION THREAD

**Discussion by dshaw619** [gitbls/sdm#68] [Q&A] [answered]
I am encountering a new error message saying "qemu-arm-static missing", but it does exist and is up to date on my Linux Mint 21 laptop.  This seems to occur at the beginning of --customize Phase 1 (I reverted to sdm-plugin-template.sh to make sure it wasn't something I had added to the plugin).  I believe this started after installing v7.7.1.  

```
...
* Enter image '/home/doug/Downloads/bullseye-armhf-customized.img' for Phase 1
? qemu-arm-static missing for chroot; try 'sudo apt install qemu-user-static binfmt-support'
umount: /mnt/sdm/boot unmounted
umount: /mnt/sdm unmounted

Thu Feb  2 05:35:05 PM PST 2023

doug@dv6-Notebook-PC:~$ sudo apt install qemu-user-static binfmt-support
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
binfmt-support is already the newest version (2.2.1-2).
qemu-user-static is already the newest version (1:6.2+dfsg-2ubuntu6.6).
The following packages were automatically installed and are no longer required:
  linux-headers-5.15.0-53 linux-headers-5.15.0-56
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 18 not upgraded.
```

**Comment by gitbls** [comment, marked as answer]
Just to confirm: Are you trying to customize a 32-bit armhf RasPiOS?

sdm looks for qemu-arm-static in /usr/bin. Can you confirm that it's there on your Mint system? If it's not there, please locate it and let me know where it's hiding.  And if it's nowhere to be found, then try `sudo apt install --reinstall qemu-user-static binfmt-support` retry, and LMK what you found.

The changes in 7.7.1 shouldn't have affected this, but I'll probaby find out I was wrong 🙄

**Reply by dshaw619** [reply]
Sure enough, qemu-arm-static wasn't in /usr/bin nor anywhere else.  
Before reinstalling:

```
doug@dv6-Notebook-PC:~$ ls -l /usr/bin/qemu-arm*
-rwxr-xr-x 1 root root 4395552 Dec  8 01:17 /usr/bin/qemu-armeb-static
```
After reinstalling qemu-arm-static and binfmt-support:

```
doug@dv6-Notebook-PC:~$ ls -l /usr/bin/qemu-arm*
-rwxr-xr-x 1 root root 4395552 Dec  8 01:17 /usr/bin/qemu-armeb-static
-rwxr-xr-x 1 root root 4383264 Dec  8 01:17 /usr/bin/qemu-arm-static
```

Don't know what might have happened to those files.  sdm --customize runs to normal completion now.

Thanks!

**Reply by dshaw619** [reply]
Forgot to answer your first question.  Yes, I am customizing a 32-bit armhf RasPiOS image.

**Comment by gitbls** [comment]
I can see one possible way that sdm might have deleted that file, but can't explain how that could actually happen. But, will tie it off in the next update.

If you happent to figure out a scenario where sdm deletes it, I'd definitely like to know how to repro it.

**Reply by 1stcall** [reply]
I had a similar problem while using pi-gen to build images. Part of the cleanup process in pi-gen, removes qemu-arm-static binary, even if it is legitimately installed.

**Reply by gitbls** [reply]
It may be that there was a similar issue with pi-gen, but the root causes of the issues are not related at all. 

If sdm is doing this, I consider it to be a bug and will get it fixed.

**Comment by dshaw619** [comment]
After the successful --customize operation yesterday, /usr/bin/qemu-arm-static was still there.  I verified that this morning and then ran a --burn.  Failed (possibly while setting up my new username) and it's no longer there.

```
doug@dv6-Notebook-PC:~$ ls -l /usr/bin/qemu-arm*
-rwxr-xr-x 1 root root 4395552 Dec  8 01:17 /usr/bin/qemu-armeb-static
-rwxr-xr-x 1 root root 4383264 Dec  8 01:17 /usr/bin/qemu-arm-static
doug@dv6-Notebook-PC:~$ 
doug@dv6-Notebook-PC:~$ ~/Documents/Briefcase/It/Raspberry_Pi/Setup/11-Bullseye/sdm-burn.sh

Fri Feb  3 10:50:00 AM PST 2023

*Beginning sdm burn

[sudo] password for doug:             
* Burn '/home/doug/Downloads/bullseye-armhf-customized.img' to '/dev/sdd'...
dd if=/home/doug/Downloads/bullseye-armhf-customized.img of=/dev/sdd status=progress bs=16M
5318377472 bytes (5.3 GB, 5.0 GiB) copied, 251 s, 21.2 MB/s
318+0 records in
318+0 records out
5335154688 bytes (5.3 GB, 5.0 GiB) copied, 270.003 s, 19.8 MB/s
> Expand Root: Expand partition from 4.7GB (5.1GiB) to 14.6GB (15.7GiB)
Checking that no-one is using this disk right now ... OK

Disk /dev/sdd: 14.84 GiB, 15931539456 bytes, 31116288 sectors
Disk model: MassStorageClass
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x1d124286

Old situation:

Device     Boot  Start      End Sectors  Size Id Type
/dev/sdd1         8192   532479  524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 10419921 9887442  4.7G 83 Linux

/dev/sdd2: 
New situation:
Disklabel type: dos
Disk identifier: 0x1d124286

Device     Boot  Start      End  Sectors  Size Id Type
/dev/sdd1         8192   532479   524288  256M  c W95 FAT32 (LBA)
/dev/sdd2       532480 31116287 30583808 14.6G 83 Linux

The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.
> Set new disk ID '9de6d13f' on '/dev/sdd'
Disk identifier changed from 0x1d124286 to 0x9de6d13f.

The partition table has been altered.
Calling ioctl() to re-read partition table.
Syncing disks.
> Disk ID 9de6d13f set successfully with sfdisk
mount: /dev/sdd2 mounted on /mnt/sdm.
mount: /dev/sdd1 mounted on /mnt/sdm/boot.
> Set new Disk ID '9de6d13f' in /etc/fstab and /boot/cmdline.txt
> Flush accumulated burn log messages
% Configure chroot to use this IMG on your ' ARM' computer
  * sdm will run more slowly due to qemu emulation
> RasPiOS firstboot service not disabled; at least one of --expand-root and --regen-ssh-host-keys not set
> Set hostname 'KeyLime'
> First System Boot Custom Boot Scripts disabled
> First System Boot automatic restart enabled
** Enter password for user 'des': 
> Run Plugins Phase '0'
> Run Plugin 'vnc' (/mnt/sdm/usr/local/sdm/plugins/vnc) Phase 0 with arguments: 
  'realvnc=1280x720'
* Plugin vnc: Start Phase 0
* Plugin vnc: Complete Phase 0
> Change passwords or add user per --user, --password-pi, --password-user, and/or 
  --password-root
> Run script 'ucp-run-script' [/etc/sdm/assets/ucp-run-script] in nspawn
> Add user 'des' to sudoers.d
cp: '/usr/bin/qemu-arm-static' and '/usr/bin/qemu-arm-static' are the same file
chroot: cannot change root directory to '/usr/bin/qemu-arm-static': Not a directory
umount: /dev/pts: target is busy.
umount: /dev: target is busy.
? qemu-arm-static missing for chroot; try 'sudo apt install qemu-user-static binfmt-support'
umount: /mnt/sdm/boot unmounted
umount: /mnt/sdm unmounted

Fri Feb  3 11:00:16 AM PST 2023

doug@dv6-Notebook-PC:~$ ls -l /usr/bin/qemu-arm*
-rwxr-xr-x 1 root root 4395552 Dec  8 01:17 /usr/bin/qemu-armeb-static
doug@dv6-Notebook-PC:~$ 
```
For full disclosure, here's the --burn command from the sdm-burn.sh script:

```
sudo /usr/local/sdm/sdm --burn /dev/sdd \
--ddsw bs=16M \
--hostname KeyLime \
--user des \
--redact \
--plugin vnc:"realvnc=1280x720" \
--expand-root \
~/Downloads/bullseye-armhf-customized.img
```

**Reply by gitbls** [reply]
Aha...thanks! I understand what's going on, will publish a fix for this late today or tomorrow.

**Comment by gitbls** [comment]
This is corrected in V7.8. Please validate. Thanks.

**Reply by dshaw619** [reply]
I reinstalled qemu-arm-static and binfmt-support on my Linux Mint laptop and then reinstalled sdm (via the Quick Start instructions in README.md).  I've run several --customize and a couple of --burn operations successfully.  Booted the burned uSD card in RPi02W successfully.

**Reply by gitbls** [reply]
Cool! Thanks for letting me know. I mistakenly decided to debug this on my "production" x64 Debian server, and ended up rebooting once and re-installing qemu and binfmt. Moved over to a test system after that and finished debugging/testing 🤣

**Reply by dshaw619** [reply]
I unmarked your previous reply re. v7.7.1 as an answer, intending to mark the v7.8 reply, but don't see a way to do that on Github.  Sigh.  Guess I'll go back and remark the previous one.
